File: 백준/Gold/1074. Z/Z.py
import sys
input = sys.stdin.readline

n, r, c = map(int, input().split())
answer = 0


def solve(l, x, y):
    global answer
    if l < 0:
        return

    if x < (2**l) and y < (2**l):  # 제 1사분면
        answer += 0
    elif x < (2**l) and y >= (2**l):  # 제 2사분면
        answer += (2**l) * (2**l)
        y -= (2**l)
    elif x >= (2**l) and y < (2**l):  # 제 3사분면
        answer += (2**l) * (2**l) * 2
        x -= (2**l)
    if x >= (2**l) and y >= (2**l):  # 제 4사분면
        answer += (2**l) * (2**l) * 3
        x -= (2**l)
        y -= (2**l)
    solve(l-1, x, y)


solve(n-1, r, c)
print(answer)

File: 백준/Gold/1074. Z/Z.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [N, r, c] = input[0].split(" ").map(Number)
let answer = 0
const solve = (n, r, c) => {
  if (n < 0) {
    return
  }
  const currentLength = 2 ** n
  if (r < currentLength && c < currentLength) {
    answer += 0
  } else if (r < currentLength && c >= currentLength) {
    answer += currentLength * currentLength * 1
    c -= currentLength
  } else if (r >= currentLength && c < currentLength) {
    answer += currentLength * currentLength * 2
    r -= currentLength
  } else {
    answer += currentLength * currentLength * 3
    r -= currentLength
    c -= currentLength
  }
  solve(n - 1, r, c)
}
solve(N - 1, r, c)
console.log(answer)